wukong/request.py

```diff
@@ -27,105 +31,135 @@ class SolrRequest(object):
+                self.current_hosts = self.master_hosts  # backward compat
+                logger.info(
+                    'Got solr nodes from zookeeper: %s',
+                    ','.join(self.master_hosts)
+                )
+                if not self.master_hosts:
+                    logger.error('Unable to find any solr nodes to make requests to')
+                    raise SolrError("zookeeper reporting all SOLR nodes as down")
+                return True
+            except Exception as e:
+                logger.exception('Failing to retrieve new SOLR hosts from zookeeper')
+        return False
+
+    def request(self, path, params, method, body=None, headers=None, is_retry=False):
         """
         Prepare data and send request to SOLR servers
         """
-        def handle_error():
-            # if the sdk knows where zookeeper lives
-            if self.zookeeper is not None:
-                self.last_error = time.time()
-                # get the current list of active nodes from zookeeper
-                curr_hosts = self.zookeeper.get_active_hosts()
-                if len(curr_hosts) == 0:
-                    # all nodes are down - so raise an error
-                    raise SolrError("SOLR reporting all nodes as down")
-                else:
-                    # if all nodes are not down, update the current_hosts list
-                    # and the servers list
-                    self.current_hosts = curr_hosts
-                    self.servers = curr_hosts
-
-            host = self.servers.pop(0)
-            return make_request(host, path)
-
-        headers = {'content-type': 'application/json'}
-        resultparams = {'wt': 'json',
-                        'omitHeader': 'true',
-                        'json.nl': 'map'}
-
-        if params is not None:
-            resultparams.update(params)
-
-        # if there hasn't been an error in 5 minutes, reset the solr_hosts
-        if self.last_error is not None and \
-                (time.time() - self.last_error) % 60 > self.check_hosts:
-
-            self.current_hosts = self.master_hosts
-            self.last_error = None
-
-        def make_request(host, path):
-            fullpath = urljoin(host, path)
+        request_headers = {
+            'content-type': 'application/json',
+        }
+        if headers:
+            request_headers.update(headers)
+
+        # Refresh our list of hosts
+        if (self.zookeeper and
+            not is_retry and
+            self._last_request and
+            ((time.time() - self._last_request) / 60) > self.REFRESH_FREQUENCY):
+            self.attempt_zookeeper_refresh()
+
+        request_params = {
+            'wt': 'json',
+            'omitHeader': True,
+            'json.nl': 'map'
+        }
+        if params:
+            request_params.update(params)
+
+        response = None
+        for host in random.sample(self.master_hosts, len(self.master_hosts)):
```

Comment:
what's the purpose of having this be shuffled? Does it matter?

Reply:
Just so that we don't always hit the same box, thus perhaps obfuscating health issues in other boxes. Also a poor person's round-robbin the load.